service_config.go

```diff
@@ -156,6 +157,20 @@ func min(a, b *int) *int {
 }
 
 func getMaxSize(mcMax, doptMax *int, defaultVal int) *int {
+	res := getRawMaxSize(mcMax, doptMax, defaultVal)
+
+	// Cap the max size to the max int of current machine due to limit of
+	// slice length.
+	res = min(res, newInt(int((^uint(0) >> 1))))
```

Comment:
For clarity, how about this:

```go
maxInt := int(^uint(0) >> 1) // the maximum signed integer value on this system
res = min(res, &maxInt)
```

(Note that your version has an extra pair of parentheses which is hard to see because of all the nesting.)

Why does `min()` take pointers?  That just complicates things further.  `maxInt` could and should be a const instead.

How about make `min()` and `get*MaxSize()` take and return ints, and deal with pointers only where you call them:

```go
c.maxSendMessageSize = newInt(getMaxSize(*mc.MaxReqSize, *c.maxSendMessageSize, defaultClientMaxSendMessageSize))
```

Reply:
I agree that we should make maxInt const.

getMaxSize was created to handle pointers to avoid dealing with nil condition in caller. It hides the complexity from caller. And min() is just to keep logic simple as we don't need to convert int to *int when returned from it. Probably we should rename these two functions to avoid future confusions (like the input and output are all pointer type)? They are created very specifically for service config and not intended for general usage.

Reply:
How about this:

- Change `callInfo.max*MessageSize` to non-pointer types, since they must always be set.  This removes [some paranoid checking](https://github.com/grpc/grpc-go/blob/5c3d956e1876656031268e248029b2c9bed7e908/stream.go#L368)
- Change `min()` to accept and return `int` instead of `*int`
- Change `get[Raw]MaxSize` to return `int`s instead of `*int`s

Is that possible? I think this would make things simpler and keep `getMaxSize` able to handle `*int` inputs.

Reply:
I don't think we can. `callInfo.max*MessageSize` is a pointer type because it will be used to receive the value set by user API `MaxCall*MsgSize` when we configure the call. And we need nil value to indicate that user doesn't set it. And then during invoke, we will reassign its value based on service config value, user API value and default value.

Reply:
OK.. well let's just make `min` deal with `int`s then?  We can deal with the pointers where it's called.  Then we can have a const for `maxInt` and use it without creating memory.

Let's make sure we add some comments on these CallOptions/ServerOptions that specify the maximum effective value is MaxUint32 due to protocol limitations.